src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -427,10 +427,9 @@ private BoundLambda ReallyBind(NamedTypeSymbol delegateType)
             MethodSymbol cacheKey = GetCacheKey(delegateType);
 
             BoundLambda returnInferenceLambda;
-            if (_returnInferenceCache.TryGetValue(cacheKey, out returnInferenceLambda) && returnInferenceLambda.InferredFromSingleType)
+            if (_returnInferenceCache.TryGetValue(cacheKey, out returnInferenceLambda) && returnInferenceLambda.InferredFromSingleType && returnInferenceLambda.Symbol.ReturnType == returnType)
             {
                 lambdaSymbol = returnInferenceLambda.Symbol;
-                Debug.Assert(lambdaSymbol.ReturnType == returnType);
                 Debug.Assert(lambdaSymbol.RefKind == refKind);
```

Comment:
Consider moving the `returnType` check here, and check `refKind` as well (to revert this part of 629e186). (I suspect we need to ensure `refKind` is the same although I haven't found a test case for that yet.)
```
if (lambdaSym.ReturnType == returnType && lambdaSym.RefKind == refKind)  
{
    ...
    goto haveLambdaBodyAndBinders;
}
```